### PR TITLE
Extract Shoot namespaces deletion as flow task

### PR DESCRIPTION
**Pull request by ialidzhikov**
**What this PR does / why we need it**:
Currently the deletion flow fails after a timeout when resources such as secrets, configMaps with finalizers exist in the Shoot namespaces (and cannot be cleaned up).
The PR extracts the Shoot namespace deletion as separate flow task that runs after `Clean Kubernetes resources` and this flow task does not fail anymore if there are remaining namespaces after the timeout. This is okay because resources as secrets, configMaps don't have any infrastructure foot print.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```noteworthy user
The shoot deletion flow now forcefully deletes all namespaces and doesn't block the flow any longer. Please note that this may leave orphaned resources in the infrastructure if you delete the shoot without deleting these infrastructure resources beforehand.
```

### CODE REVIEW

**Review by rfranzke — DISMISSED** [review]
/lgtm, nice!

**Comment by adracus** [comment]
I see you're trying to workaround the fact that namespaces cannot be properly finalized by ignoring a timeout issue - rather than that, I'd propose creating another - for namespaces-only - implementation of the `cleanResourceFn` that tries to call the `CoreV1().Namespaces().FinalizeNamespace()` to finalize a namespace. That way, we 
* don't have to wait for the full timeout
* don't have to use `context.DeadlineExceeded` for detecting that a namespace couldn't be finalized.

**Comment by ialidzhikov** [comment]
Thanks @adracus for the suggestion. +1

Ok, looks like controller-runtime client cannot update subresources atm - kubernetes-sigs/controller-runtime#172. :( 

My initial approach was to add
```golang
type FinalizerFunc func(ctx context.Context, c client.Client, obj runtime.Object) error
```

to `CleanOptions` and to have `DefaultFinalizerFunc` (does patch for .meta.finalizers) and `NamespaceFinalizerFunc` (does patch for .meta.finalizers and .spec.finalizers) but yeah - it should be the finalize subresource `PUT to /api/v1/namespaces/{name}/finalize` to make it work - https://github.com/kubernetes/kubernetes/issues/77086#issuecomment-486832745.

I will check how I can use `CoreV1().Namespaces().Finalize(ns)` from client-go.

**Comment by ialidzhikov** [comment]
Having a separate `cleanResourceFn` for namespaces would lead to code duplication. 
Introduced the namespace finalization in the `Cleaner`. WDYT?

File: pkg/operation/botanist/cleanup.go [hunk 1]
```diff
@@ -214,3 +212,12 @@ func (b *Botanist) CleanKubernetesResources(ctx context.Context) error {
 		cleanResourceFn(c, &corev1.PersistentVolumeClaimList{}, PersistentVolumeClaimCleanOptions, ZeroGracePeriod, FinalizeAfterFiveMinutes),
 	)(ctx)
 }
+
+// CleanShootNamespaces deletes all non-system namespaces in the Shoot cluster.
+// It assumes that all workload resources are cleaned up in previous step(s).
+// If a context timeout error occurs, it gives up waiting the remaining namespaces to be cleaned up.
+func (b *Botanist) CleanShootNamespaces(ctx context.Context) error {
```
Comment (adracus) [inline pkg/operation/botanist/cleanup.go:226]: Previously, we did this simultaneously - why do it differently now?

File: pkg/utils/kubernetes/client/client.go [hunk 2]
```diff
@@ -239,17 +242,17 @@ func DefaultCleaner() Cleaner {
 }
 
 // Clean deletes and optionally finalizes resources that expired their termination date.
-func (o *cleaner) Clean(ctx context.Context, c client.Client, obj runtime.Object, opts ...CleanOptionFunc) error {
+func (o *cleaner) Clean(ctx context.Context, k8sInterface kubernetes.Interface, obj runtime.Object, opts ...CleanOptionFunc) error {
```
Comment (adracus) [inline pkg/utils/kubernetes/client/client.go]: Please don't change the interface - rather either
* create a new implementation of the `Cleaner` interface that supports cleaning a namespace
-> Here, you could re-use the existing `cleaner`. Once you have this, it's easy to re-use the existing `CleanOps`.
* write a single function that handles finalizing namespaces and have this completely separate from the generic `Cleaner`

File: pkg/utils/kubernetes/client/client.go [hunk 3]
```diff
@@ -277,11 +285,58 @@ func (o *cleaner) doClean(ctx context.Context, c client.Client, obj runtime.Obje
...
+		return err
+	}
+
+	withFinalizers := obj.DeepCopyObject()
+	acc.SetFinalizers(nil)
+	return k8sInterface.Client().Patch(ctx, obj, client.MergeFrom(withFinalizers))
+}
+
+func finalizeNamespace(ctx context.Context, k8sInterface kubernetes.Interface, obj runtime.Object) error {
+	namespace, ok := obj.(*corev1.Namespace)
```
Comment (adracus) [inline pkg/utils/kubernetes/client/client.go]: For a namespace, you still need to also remove the regular finalizers.

**Review by adracus — CHANGES_REQUESTED** [review]

File: pkg/operation/botanist/cleanup.go (hunk 1 above)
Comment (ialidzhikov) [reply pkg/operation/botanist/cleanup.go:226]: The initial idea was `CleanShootNamespaces` to give up when a timeout occurs.
And `CleanKubernetesResources` is ensuring that all resources that have IaaS foot print are cleaned up.

File: pkg/utils/kubernetes/client/client.go (hunk 2 above)
Comment (ialidzhikov) [reply pkg/utils/kubernetes/client/client.go]: Introduced `Finalizer` interface. Interface changes are now reverted.

File: pkg/utils/kubernetes/client/client.go (hunk 3 above)
Comment (ialidzhikov) [reply pkg/utils/kubernetes/client/client.go]: Setting both `.meta.finalizers` and `.spec.finalizers` and `CoreV1().Namespaces().Finalize(namespace)` does the trick.

**Review by rfranzke — APPROVED** [review]

File: pkg/operation/botanist/cleanup.go (hunk 1 above)
Comment (ialidzhikov) [reply pkg/operation/botanist/cleanup.go:226]: We do now properly finalize namespaces and having this flow task in the `flow.Parallel[...]` will mean we will always finalize the ns even if some `CleanAndEnsureGone` are not completed/failed.

Comment (rfranzke) [reply pkg/operation/botanist/cleanup.go:226]: +1, it's safer to do it only after all the other clean operations succeeded.

**Review by adracus — APPROVED** [review]

**Comment by adracus** [comment]
That PR is very nice!

**Pull request merged**